packages/@react-stately/grid/src/GridCollection.ts

```diff
@@ -29,86 +29,94 @@ export class GridCollection<T> implements IGridCollection<T> {
-          child.parentKey = node.key;
-        }
-        childKeys.add(child.key);
+      if (node.type === 'headerrow' || node.type === 'item' || node.type === 'section' || node.type === 'header') {
+        let defaultProps = {
+          level: 0,
+          key: 'row-' + i,
+          type: 'row',
+          value: undefined,
+          hasChildNodes: true,
+          childNodes: [...node.childNodes],
+          rendered: undefined,
+          textValue: undefined,
+          colspan: node.type === 'section' || node.props?.isSectionHeader ? this.columnCount : 1
+        };
+
+        let newProps = {
+          index: node.type === 'section' ? sectionCounter++ : rowCounter++,
+          rowIndex: node.type !== 'section' ? rowIndex : undefined
+        };
+
+        // Use Object.assign instead of spread to preseve object reference for keyMap. Also ensures retrieving nodes
+        // via .childNodes returns the same object as well
+        Object.assign(node, defaultProps, {...node}, newProps);
+      }
 
+      if (node.type !== 'cell') {
         if (last) {
-          last.nextKey = child.key;
-          child.prevKey = last.key;
-        } else {
-          child.prevKey = null;
+          last.nextKey = node.key;
+          node.prevKey = last.key;
         }
-
-        visit(child);
-        last = child;
+        last = node;
       }
 
-      if (last) {
-        last.nextKey = null;
-      }
+      this.keyMap.set(node.key, node);
 
-      // Remove deleted nodes and their children from the key map
-      if (prevNode) {
-        for (let child of prevNode.childNodes) {
-          if (!childKeys.has(child.key)) {
-            remove(child);
+      let childKeys = new Set();
+      let lastCell: GridNode<T>;
+      let index = 0;
+      for (let child of node.childNodes) {
+        if (child.type === 'cell') {
+          if (child.parentKey == null) {
+            // if child is a cell parent key isn't already established by the collection, match child node to parent row
+            child.parentKey = node.key;
+          }
+          childKeys.add(child.key);
+
+          if (lastCell) {
+            lastCell.nextKey = child.key;
+            child.prevKey = lastCell.key;
+          } else {
+            child.prevKey = null;
+          }
+
+          visit(child);
+          lastCell = child;
+        } else {
+          if (last) {
+            last.nextKey = child.key;
+            child.prevKey = last.key;
+          } else {
+            child.prevKey = null;
           }
+          visit(child, index++);
+          last = child;
         }
       }
-    };
 
-    let remove = (node: GridNode<T>) => {
-      this.keyMap.delete(node.key);
-      for (let child of node.childNodes) {
-        if (this.keyMap.get(child.key) === child) {
-          remove(child);
-        }
+      if (node.type === 'item' || node.type === 'headerrow' || node.type === 'header') {
+        this.rows.push(node);
+        rowIndex++;
```

Comment:
Recommenting:
From testing, it feels like we either keep aria-rowindex as an absolute position index (aka using this rowIndex, may not be feasible in RAC land) or use aria-posinset + aria-level + aria-setsize with relative position indicies (aka do away with rowIndex). Making aria-rowindex relative to the immediate parent makes announcements in VO confusing since it will announce values relative to the total number of rows rather than the section.